pkg/azuredx/datasource.go

```diff
@@ -62,6 +63,11 @@ func NewDatasource(settings backend.DataSourceInstanceSettings) (instancemgmt.In
 	}
 	adx.client = client.New(httpClient)
 
+	adx.serviceCredentials = azureauth.ServiceCredentials{
+		DatasourceSettings: *adx.settings,
+		PostForm:           httpClient.PostForm,
+	}
```

Comment:
I would avoid this inner `serviceCredentials` struct for two reasons: the settings are duplicated (`adx.settings` and `adx.serviceCredentials.DatasourceSettings`) and it's only necessary for the on-behalf-of scenario. I would add the settings and the httpClient as parameters for the `GetOnBehalfOfBearerToken` function instead.

Reply:
We need the settings for the service-principal and user-forwarding  options too. I have a follow up patch which completes this architecture by moving the `tokenprovider` into ServiceCredentials too. Then authorization is nicely handled in one progmatic function, rather than hidden in layers form another code base. See the TODO for details.

Reply:
not feeling strongly about this one either